[FEATURE][processing] Add custom OPTIONS to contour algorithm

gkimbell opened the pull request:
Suggested inclusion of OPTIONS parameter in gdal:contour

gkimbell committed:
Update contour.py

nyalldawson commented:
Looks good! Can you add a unit test to https://github.com/qgis/QGIS/blob/master/python/plugins/processing/tests/GdalAlgorithmsTest.py#L550 covering this too please?

nyalldawson commented:
@gkimbell It would be great to get this merged in time for QGIS 3.4, but that leaves us with just 4 days to get this merge-worthy.

Things to do are:

- Fix the error shown in https://travis-ci.org/qgis/QGIS/jobs/426057981
- Add a new unit test to https://github.com/qgis/QGIS/blob/master/python/plugins/processing/tests/GdalAlgorithmsTest.py#L550 with custom options set
- Run the scripts/prepare-commit.sh script over the changes to ensure that the auto-formatting scripts are applied

alexbruy commented:
@gkimbell why not use custom parameter and widget for this to be consistent with the rest of the GDAL algorithms?

gkimbell committed:
Update GdalAlgorithmsTest.py

First draft of a unit test for fixed level contours (L592-602). May change depending on the parameter input method used. Currently the interval (-i) parameter is always included in the command line (defaults to 10 if not specified), but is overridden by the fixed level (-fl) contour list if provided.

nyalldawson commented on python/plugins/processing/algs/gdal/contour.py:
```diff
@@ -74,6 +75,10 @@ def initAlgorithm(self, config=None):
                                                        self.tr('Attribute name (if not set, no elevation attribute is attached)'),
                                                        defaultValue='ELEV',
+                                                       optional=True))       
+        self.addParameter(QgsProcessingParameterString(self.OPTIONS,
```
This should be marked as an "advanced" option. See 

https://github.com/qgis/QGIS/blob/master/python/plugins/processing/algs/gdal/ClipVectorByMask.py#L58

You need the whole block:

        options_param = QgsProcessingParameterString(self.OPTIONS,
                                                     self.tr('Additional creation options'),
                                                     defaultValue='',
                                                     optional=True)
        options_param.setFlags(options_param.flags() | QgsProcessingParameterDefinition.FlagAdvanced)
        self.addParameter(options_param)

nyalldawson commented on python/plugins/processing/algs/gdal/contour.py:
```diff
@@ -162,6 +167,9 @@ def getConsoleCommands(self, parameters, context, feedback, executing=True):
 
         if offset:
             arguments.append('-off {}'.format(offset))
+        
```
Before this, you'll need to retrieve the options value. See https://github.com/qgis/QGIS/blob/master/python/plugins/processing/algs/gdal/ClipVectorByMask.py#L86

    options = self.parameterAsString(parameters, self.OPTIONS, context)
    if options:
      ...

gkimbell replied:
That's great Nyall - I've included your suggestions in the version in my fork (and they work fine in a local copy)
I've also added a unit test to GdalAlgorithmsTest.py in my fork (lines 592-603)
What's the next step - do I have to resubmit the pull request?
Thanks for your patience!
Geoff

nyalldawson replied:
Nothing more for you to do, until the Travis CI results come in. If they are all green I'll merge, otherwise I'll advise what the next step is.

gkimbell committed:
Update contour.py

nyalldawson commented:
Looks good! Let's see what Travis says...

gkimbell committed:
Update GdalAlgorithmsTest.py

nyalldawson commented on python/plugins/processing/tests/GdalAlgorithmsTest.py:
```diff
@@ -589,6 +589,18 @@ def testContour(self):
+                                    'BAND': 1,
+                                    'FIELD_NAME': 'elev',
+                                    'INTERVAL': 0,
+                                    'OPTIONS': '-fl 100 125 150 200'
```
Missing a ',' on the end of this line!

gkimbell committed:
Update GdalAlgorithmsTest.py

gkimbell committed:
Update contour.py

nyalldawson commented:
Thanks for all your work here!

gkimbell commented:
Sorry it took so long - I'll familiarise myself with the auto-formatting scripts before the next time!

nyalldawson commented:
Not a problem -- there is a steep learning curve here trying to get familiar with the different infrastructure and requirements in place. 

I'd encourage a read over https://docs.qgis.org/testing/en/docs/developers_guide/ . While a lot is C++ oriented, there's still useful info about Travis and tests, etc.

Just ask on the qgis developer mailing list if you have questions -- there's a lot of people who are happy to help!
